### DISCUSSION TITLE
Is there a "disallow foreign attributes" config option?

### DISCUSSION THREAD

**Discussion by dan-fritchman** [pydantic/pydantic#6232] [Question]
Example: 

```python
from pydantic.dataclasses import dataclass
from pydantic import Extra

class Config:
    extra = Extra.forbid
    validate_assignment = True

@dataclass(config=Config)
class C:
    a: int

c = C(a=1)

# I expected this to fail, but it doesn't
c.q = 11

# This fails though
c.a = TabError
```

Is there a configuration option that would cause the assignment to `c.q` to fail? 

One thing that does make `C.q` assignment fail is `allow_mutation=False` (and/or the dataclasses `frozen` option). But in many cases I *do* want to mutate the "real fields" (e.g. `C.a`). 

Thanks!

**Comment by gyanu2507** [comment]
`extra = Extra.forbid` only applies when the object is constructed (or when assignment goes through Pydantic validation). A Pydantic dataclass is still a normal dataclass afterward, so `c.q = 11` just sets a Python attribute.

If you want unknown assignment to fail and still mutate `a`, use a `BaseModel`:

```python
from pydantic import BaseModel, ConfigDict

class C(BaseModel):
    model_config = ConfigDict(extra="forbid", validate_assignment=True)
    a: int

c = C(a=1)
c.a = 2   # ok
c.q = 11  # ValidationError
```

On a dataclass, `frozen=True` blocks every assignment, including `a`. There is no "forbid extra attrs, allow real fields" switch for pydantic dataclasses that I know of.

**Comment by nanisadw3** [comment]
Hi @dan-fritchman 👋

In Python, standard dataclasses allow dynamic attribute assignment by default unless you use `slots`. If you are using dataclasses with Pydantic, you can prevent adding arbitrary new attributes after instantiation by enabling `slots=True`:

```python
from pydantic.dataclasses import dataclass
from pydantic import ConfigDict

@dataclass(config=ConfigDict(validate_assignment=True), slots=True)
class C:
    a: int

c = C(a=1)

# This will now fail as expected:
# AttributeError: 'C' object has no attribute 'q'
c.q = 11 
```

Alternatively, if you use standard Pydantic `BaseModel`, you can configure `extra='forbid'` along with `validate_assignment=True`:

```python
from pydantic import BaseModel, ConfigDict, ValidationError

class C(BaseModel):
    model_config = ConfigDict(extra='forbid', validate_assignment=True)
    a: int

c = C(a=1)

try:
    c.q = 11
except (ValueError, ValidationError) as e:
    print("Blocked adding foreign attribute:", e)
```

Using either `slots=True` on the dataclass or `extra='forbid'` on `BaseModel` completely stops foreign attributes from being attached to instances.
